### ISSUE TITLE
memory crash with umlauts or special characters 

### ISSUE THREAD

**Issue by gebrerav** [gkralik/php7-sapnwrfc#101]
We migrate from Eduard Koucky's SAP Rfc library to sapnwrfc.
One of our self developed RFC funtion calls that works flawlessly with Koucky's SAPRFC, crahes if called with php7-sapnwrfc.

After some debugging we found that the crash happens if one field in the records returned by the function call  contains several umlauts.
The crash "memory corruption or free(): invalid next size (fast) " happens if the field is 40 characters long and contains more than 3 umlauts.
The structure of the table is :

AUFNR                        CHARACTER       12
KTEXT                          CHARACTER       40
POSID                          CHARACTER       24
STIWO                         CHARACTER       16
VRGNG                        CHARACTER       4
VKTXT                          CHARACTER       40
MENGE                        CHARACTER       17
MEINH                         CHARACTER       3
RUECK                          NUMC            10
ARBPL                          CHARACTER       8
SSAVD                          DATE

The error was reproducable with field KTEXT and the function returning  one record.
in a further attempt to debug the error, i called standard SAP function call 'STFC_DEEP_TABLE'
with only one record:
$parms = ['IMPORT_TAB' => [['I' => 123, 'C' => 'Äöß/€üöuCu', 'STR' => 'The quick brown fox ...' ]]];
$results = $f->invoke($parms);
var_dump($results);

result of var_dump:

array(2) {
  ["EXPORT_TAB"]=>
  array(2) {
    [0]=>
    array(4) {
      ["I"]=>
      int(123)
      ["C"]=>
      **string(15) "Äöß/#üöuCu"**
      ["STR"]=>
      string(23) "The quick brown fox ..."
      ["XSTR"]=>
      string(0) ""
    }
    [1]=>
    array(4) {
      ["I"]=>
      int(10)
      ["C"]=>
      string(10) "Appended  "
      ["STR"]=>
      string(8) "20201216"
      ["XSTR"]=>
      string(4) "ޭ▒▒"
    }
  }
  ["RESPTEXT"]=>
  string(255) "SAP R/3 Rel. 740   Sysid: SHT      Date: 20201216   Time: 170901                                                                                                                                                                                               "
}

calling the same with '0123456789' in Field C returned 
..
 ["C"]=>
      **string(10) "0123456789"** instead of string(15) , apparently umlauts are converted to 16 bit that do not fit in the memory that was allocated for the 40C field.

Most probably the error will also show up with other function calls , though the probabilty to have 4 or more umlauts in a field with 40 characters is not very high..

**Comment by gkralik** [comment]
Thanks for reporting this, I will have a look in the coming days.

**Comment by gkralik** [comment]
I'm having a hard time reproducing this.

Can you share:
- your connection configuration (excl. credentials)
- a short ABAP FM that reliably reproduces the crash for you.

The string length in PHP is the byte-length of the string, so for an input of `Äöß/€üöuCu` the string length in PHP is expected to be 17 (umlauts are encoded as 2 bytes, € as 3 bytes):
```php
echo strlen('Äöß/€üöuCu');
// => 17
echo mb_strlen('Äöß/€üöuCu');
// => 10
```

Why it is 15 four you I do not yet know.

**Comment by gebrerav** [comment]
Hi Gregor,

thanks for Your fast response. I will provide a minimal ABAP to You , most probably only a kind of write to table ..
The connection information is :

$config = [
    'ashost' => '1.1.1.1',
    'sysnr'  => '00',
    'client' => '100',
    'user'   => 'sapuser',
    'passwd' => 'userspassword',
    'trace'  => SapConnection::TRACE_LEVEL_OFF,
];

Best Regards Gerhard
Von: Gregor Kralik <notifications@github.com>
Gesendet: Samstag, 19. Dezember 2020 19:00
An: gkralik/php7-sapnwrfc <php7-sapnwrfc@noreply.github.com>
Cc: Brehm Gerhard <Gerhard.Brehm@andritz.com>; Author <author@noreply.github.com>
Betreff: Re: [gkralik/php7-sapnwrfc] memory crash with umlauts or special characters (#101)


NON-ANDRITZ SOURCE: BE CAUTIOUS WITH CONTENT, LINKS OR ATTACHMENTS.

I'm having a hard time reproducing this.

Can you share:

  *   your connection configuration (excl. credentials)
  *   a short ABAP FM that reliably reproduces the crash for you.

The string length in PHP is the byte-length of the string, so for an input of Äöß/€üöuCu the string length in PHP is expected to be 17 (umlauts are encoded as 2 bytes, € as 3 bytes):

echo strlen('Äöß/€üöuCu');



// => 17



echo mb_strlen('Äöß/€üöuCu');



// => 10

Why it is 15 four you I do not yet know.

**Comment by gebrerav** [comment]
Hi Gregor
Attached is the short ABAP and the php call
Thanks and best Regards Gerhard

-------- Originalnachricht--------
Von: Gregor Kralik <notifications@github.com<mailto:notifications@github.com>>
Datum: Sa., 19. Dez. 2020, 19:00
An: gkralik/php7-sapnwrfc <php7-sapnwrfc@noreply.github.com<mailto:php7-sapnwrfc@noreply.github.com>>
CC: Brehm Gerhard <Gerhard.Brehm@andritz.com<mailto:Gerhard.Brehm@andritz.com>>,Author <author@noreply.github.com<mailto:author@noreply.github.com>>
Betreff: Re: [gkralik/php7-sapnwrfc] memory crash with umlauts or special characters  (#101)

NON-ANDRITZ SOURCE: BE CAUTIOUS WITH CONTENT, LINKS OR ATTACHMENTS.


I'm having a hard time reproducing this.

Can you share:

  *   your connection configuration (excl. credentials)
  *   a short ABAP FM that reliably reproduces the crash for you.

The string length in PHP is the byte-length of the string, so for an input of Äöß/€üöuCu the string length in PHP is expected to be 17 (umlauts are encoded as 2 bytes, € as 3 bytes):

echo strlen('Äöß/€üöuCu');

// => 17

echo mb_strlen('Äöß/€üöuCu');

// => 10

Why it is 15 four you I do not yet know.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://hes32-ctp.trendmicro.com:443/wis/clicktime/v1/query?url=https%3a%2f%2fgithub.com%2fgkralik%2fphp7%2dsapnwrfc%2fissues%2f101%23issuecomment%2d748505170&umid=f08f2a3a-319c-4539-95b0-bfc0deea5b19&auth=144056baf7302d777acad187aac74d4b9ba425e1-7ed1762903e8eb81db0aac210f5d5f23744f68a5>, or unsubscribe<https://hes32-ctp.trendmicro.com:443/wis/clicktime/v1/query?url=https%3a%2f%2fgithub.com%2fnotifications%2funsubscribe%2dauth%2fASEJ4WWJSIAW5SJESZUBCADSVTSZ3ANCNFSM4U6HQFNA&umid=f08f2a3a-319c-4539-95b0-bfc0deea5b19&auth=144056baf7302d777acad187aac74d4b9ba425e1-c0c940b2c867f7568ac2cf1b0fd02287912c7a03>.

<?php
ini_set('memory_limit','1024M');
use SAPNWRFC\Connection as SapConnection;
use SAPNWRFC\Exception as SapException;

$config = [
    'ashost' => '143.161.104.11',
    'sysnr'  => '00',
    'client' => '100',
    'user'   => 'brehm_ge',
    'passwd' => '############',
    'trace'  => SapConnection::TRACE_LEVEL_OFF,
];

$conn = new SapConnection($config);

$f = $conn->getFunction('Z_RFC_TEST');

$results = $f->invoke([
]);

print_r($results);

$conn->close();


FUNCTION Z_RFC_TEST
  TABLES
    ET_OUTPUT_TABLE LIKE BAPI_DOC_DRAT.



  "
  CLEAR et_output_table[].
  "
  DO 1 TIMES.
    INSERT INITIAL LINE INTO TABLE et_output_table[] ASSIGNING FIELD-SYMBOL(<ls_data>).
    "Values without dump
    "    <ls_data>-description = '1234567890123456789012345678901234567890'.
    "   <ls_data>-description = 'Äußeres Ölrohr Nacharbeit; '.
    "    <ls_data>-description = 'ÄßÖ;'.
    "    <ls_data>-description = 'ÄßØ'.
    "    <ls_data>-description = 'ÄØ'.
    "
    "Values with dump
    "    <ls_data>-description = 'Äußeres Ölrohr Nacharbeit; Ø245/156x3120'.
    "    <ls_data>-description = 'Äußeres Ölrohr Nacharbeit; Ø'.
    "    <ls_data>-description = 'ÄßÖ;Ø/'.
    "    <ls_data>-description = 'ÄßÖ;Ø'.
    "    <ls_data>-description = 'ÄßÖØ'.
    "    <ls_data>-description = 'ÄÖÜØ'.
    <ls_data>-description = 'ÄÖÜß'.
  ENDDO.

ENDFUNCTION.

**Comment by gkralik** [comment]
Thank you.
Can you also provide
- the PHP version you experienced this on
- the exact version of php7-sapnwrfc you used (version or git commit)

**Comment by gkralik** [comment]
And if you happen to have a coredump from one of the crashes that would greatly help with debugging.

**Comment by gebrerav** [comment]
hi

unfortunately i was not able to create a coredump, so i called the script via gdb and with strace.


Output of gdb php > run t,php

warning: Loadable section ".note.gnu.property" outside of ELF segments
...
warning: Loadable section ".note.gnu.property" outside of ELF segments
free(): invalid next size (fast)
Program received signal SIGABRT, Aborted.
0x00007ffff4b7b70f in raise () from /lib64/libc.so.6



Output of strace -o strace.out php t,php is attached in file strace.php


regards gerhard



________________________________
Von: Gregor Kralik <notifications@github.com>
Gesendet: Dienstag, 22. Dezember 2020 17:01
An: gkralik/php7-sapnwrfc
Cc: Brehm Gerhard; Author
Betreff: Re: [gkralik/php7-sapnwrfc] memory crash with umlauts or special characters (#101)


NON-ANDRITZ SOURCE: BE CAUTIOUS WITH CONTENT, LINKS OR ATTACHMENTS.


And if you happen to have a coredump from one of the crashes that would greatly help with debugging.

**Comment by gkralik** [comment]
@gebrerav Unfortunately, GitHub seems to strip attachments. Please attach the files to this issue instead.

**Comment by gebrerav** [comment]
Hi Gregor,
i uploaded the strace file as well as the php and abap.

is it also possible to hire You to solve the problem ? 
If yes, and You are interested, please contact me via gerhard.brehm@brehm-rv.de
Thanks in advance
Gerhard 

[strace.txt](https://github.com/gkralik/php7-sapnwrfc/files/5759111/strace.txt)
[php.txt](https://github.com/gkralik/php7-sapnwrfc/files/5759113/php.txt)
[abap.txt](https://github.com/gkralik/php7-sapnwrfc/files/5759114/abap.txt)

**Comment by gkralik** [comment]
@gebrerav You have closed the issue. Is the problem resolved?

I am still unable to reproduce this problem using the latest version of *php7-sapnwrfc* (built from source), SAP NW RFC SDK 7.50 (patch level 5) and PHP 7.4.13.

The *strace* output you provided indicates a core dump has been created (`+++ killed by SIGABRT (core dumped) +++`). To further investigate, I need that core dump (use `ulimit -c 0` before executing the test script - if you are using *systemd* you should be able to extract it using `coredumpctl`).

Additionally, this report still lacks the following information:
- the exact PHP version you are using
- the exact version of *php7-sapnwrfc* you used (version or git commit)
- the SAP NW RFC SDK version you are using

PS: Donations are always welcome if you find the extension useful - you can find the PayPal link in the README.

**Comment by gebrerav** [comment]
Good Morning,
Apparently i have closed the issue by mistake,the problem isnt resolved.
As I'm on holiday actually, i will not be able to provide the information about php /SAP rfc version before next week.
Regards Gerhard

-------- Originalnachricht--------
Von: Gregor Kralik <notifications@github.com<mailto:notifications@github.com>>
Datum: So., 3. Jan. 2021, 18:01
An: gkralik/php7-sapnwrfc <php7-sapnwrfc@noreply.github.com<mailto:php7-sapnwrfc@noreply.github.com>>
CC: Brehm Gerhard <Gerhard.Brehm@andritz.com<mailto:Gerhard.Brehm@andritz.com>>,Mention <mention@noreply.github.com<mailto:mention@noreply.github.com>>
Betreff: Re: [gkralik/php7-sapnwrfc] memory crash with umlauts or special characters  (#101)

NON-ANDRITZ SOURCE: BE CAUTIOUS WITH CONTENT, LINKS OR ATTACHMENTS.


@gebrerav<https://hes32-ctp.trendmicro.com:443/wis/clicktime/v1/query?url=https%3a%2f%2fgithub.com%2fgebrerav&umid=99209443-155b-43e6-a022-d2d5972658b1&auth=144056baf7302d777acad187aac74d4b9ba425e1-1127def10e9ce1ddeb7ce214fdfcee0231cee612> You have closed the issue. Is the problem resolved?

I am still unable to reproduce this problem using the latest version of php7-sapnwrfc (built from source), SAP NW RFC SDK 7.50 (patch level 5) and PHP 7.4.13.

The strace output you provided indicates a core dump has been created (+++ killed by SIGABRT (core dumped) +++). To further investigate, I need that core dump (use ulimit -c 0 before executing the test script - if you are using systemd you should be able to extract it using coredumpctl).

Additionally, this report still lacks the following information:

  *   the exact PHP version you are using
  *   the exact version of php7-sapnwrfc you used (version or git commit)
  *   the SAP NW RFC SDK version you are using

PS: Donations are always welcome if you find the extension useful - you can find the PayPal link in the README.

**Comment by gebrerav** [comment]
Hi,


I will upload the coredump to the issue in a few minutes.


PHP Version is :

PHP 7.4.11 (cli) (built: Sep 29 2020 10:17:06) ( NTS )
Copyright (c) The PHP Group
Zend Engine v3.4.0, Copyright (c) Zend Technologies
   with Zend OPcache v7.4.11, Copyright (c), by Zend Technologies

SAPNW Library : 7.50.7


php7-sapnwrfc : Build from source , version 1.3.0 -2019-02-23 , https://github.com/gkralik/php7-sapnwrfc/archive/main.zip<https://gi>


regards Gerhard



<https://github.com/gkralik/php7-sapnwrfc/releases/tag/1.3.0>







________________________________
Von: Gregor Kralik <notifications@github.com>
Gesendet: Sonntag, 3. Januar 2021 18:01
An: gkralik/php7-sapnwrfc
Cc: Brehm Gerhard; Mention
Betreff: Re: [gkralik/php7-sapnwrfc] memory crash with umlauts or special characters (#101)


NON-ANDRITZ SOURCE: BE CAUTIOUS WITH CONTENT, LINKS OR ATTACHMENTS.


@gebrerav<https://hes32-ctp.trendmicro.com:443/wis/clicktime/v1/query?url=https%3a%2f%2fgithub.com%2fgebrerav&umid=99209443-155b-43e6-a022-d2d5972658b1&auth=144056baf7302d777acad187aac74d4b9ba425e1-1127def10e9ce1ddeb7ce214fdfcee0231cee612> You have closed the issue. Is the problem resolved?

I am still unable to reproduce this problem using the latest version of php7-sapnwrfc (built from source), SAP NW RFC SDK 7.50 (patch level 5) and PHP 7.4.13.

The strace output you provided indicates a core dump has been created (+++ killed by SIGABRT (core dumped) +++). To further investigate, I need that core dump (use ulimit -c 0 before executing the test script - if you are using systemd you should be able to extract it using coredumpctl).

Additionally, this report still lacks the following information:

  *   the exact PHP version you are using
  *   the exact version of php7-sapnwrfc you used (version or git commit)
  *   the SAP NW RFC SDK version you are using

PS: Donations are always welcome if you find the extension useful - you can find the PayPal link in the README.

**Comment by gebrerav** [comment]
corefile attached
[core.zip](https://github.com/gkralik/php7-sapnwrfc/files/5775865/core.zip)

**Comment by gkralik** [comment]
Thanks for providing more detailed information.

While I could not reproduce your specific issue, I found a memory corruption in the code responsible for converting strings from SAP to PHP and vice versa that looks similar to your issue.

I have pushed a possible fix in dfc332c.

Using PHP 7.3.25, 7.4.11 and 7.4.13 with both SDK 7.50 patch level 5 and 7 I could not trigger the memory corruption again after applying the fix.

Can you build the extension again with that fix and see if your issue still persists?

**Comment by gebrerav** [comment]
Hi,

Thanks for providing the fix. Apparently, the issued does not occur after building the extension with the fix and testing it as well with the test function call and also the original application.
The issue is solved


Von: Gregor Kralik <notifications@github.com>
Gesendet: Samstag, 9. Januar 2021 13:53
An: gkralik/php7-sapnwrfc <php7-sapnwrfc@noreply.github.com>
Cc: Brehm Gerhard <Gerhard.Brehm@andritz.com>; Mention <mention@noreply.github.com>
Betreff: Re: [gkralik/php7-sapnwrfc] memory crash with umlauts or special characters (#101)


NON-ANDRITZ SOURCE: BE CAUTIOUS WITH CONTENT, LINKS OR ATTACHMENTS.

Thanks for providing more detailed information.

While I could not reproduce your specific issue, I found a memory corruption in the code responsible for converting strings from SAP to PHP and vice versa that looks similar to your issue.

I have pushed a possible fix in dfc332c<https://hes32-ctp.trendmicro.com:443/wis/clicktime/v1/query?url=https%3a%2f%2fgithub.com%2fgkralik%2fphp7%2dsapnwrfc%2fcommit%2fdfc332c624e5591fb916b9f512e7e17348fb8637&umid=1dae8ef0-8b7e-4619-97f4-146bea1bf221&auth=144056baf7302d777acad187aac74d4b9ba425e1-828019e3721a32fdd176ba6c1d4a13212dd7a5ee>.

Using PHP 7.3.25, 7.4.11 and 7.4.13 with both SDK 7.50 patch level 5 and 7 I could not trigger the memory corruption again after applying the fix.

Can you build the extension again with that fix and see if your issue still persists?

**Comment by gkralik** [comment]
The fix is included in release [1.4.0](https://github.com/gkralik/php7-sapnwrfc/releases/tag/1.4.0).
